lib/posix/pthread.c

```diff
@@ -254,31 +410,32 @@ int pthread_setcancelstate(int state, int *oldstate)
  */
 int pthread_cancel(pthread_t pthread)
 {
-	struct posix_thread *thread = to_posix_thread(pthread);
+	void *retval;
+	int detachstate;
 	int cancel_state;
-	k_spinlock_key_t cancel_key;
+	k_spinlock_key_t key;
+	struct posix_thread *t;
 
-	if ((thread == NULL) || (thread->state == PTHREAD_TERMINATED)) {
+	t = to_posix_thread(pthread);
+	if (t == NULL) {
 		return ESRCH;
 	}
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_pending = 1;
-	cancel_state = thread->cancel_state;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
-
+	key = k_spin_lock(&t->lock);
+	t->cancel_pending = true;
```

Comment:
Should this check to see if `cancel_pending` is already `true`?

Reply:
in the previous implementation it was using the same logic. The only pace where `cancel_pending` is checked from what I've seen is `pthread_setcancelstate()`